### PR TITLE
ENH: cluster: add an optional argument `seed` for `kmeans` and `kmeans2` to set random generator and random state.

### LINKED ISSUE #9450: allow seeding of init methods in vq.kmeans2
The `vq.kmeans2` function has initialization methods (that can be selected by the `minit` keyword) that do random number generation without allowing for seeding. See e.g. https://github.com/scipy/scipy/pull/9037#pullrequestreview-137675449

It should be possible for the user to pass in a `np.random.RandomState` instance or set a seed, rather than having to rely on setting the global state. Can be done in a similar way as the `random_state` keyword in the stats distributions.

### PR DESCRIPTION
**Pull request by AtsushiSakai**
#### Reference issue

fix #9450

#### What does this implement/fix?

I added an optional argument `seed` for `kmeans` and `kmeans2` to set random generator and random state.
I also added tests for these and updated docs.

### CODE REVIEW

File: scipy/cluster/vq.py [hunk 1]
```diff
@@ -314,7 +314,8 @@ def _kmeans(obs, guess, thresh=1e-5):
     return code_book, prev_avg_dists[1]
 
 
-def kmeans(obs, k_or_guess, iter=20, thresh=1e-5, check_finite=True):
+def kmeans(obs, k_or_guess, iter=20, thresh=1e-5, check_finite=True,
+           seed=None):
```
Comment (tupui) [inline scipy/cluster/vq.py:318]: Now we advice to mark these as keyword arguments only #13955
```suggestion
def kmeans(obs, k_or_guess, iter=20, thresh=1e-5, check_finite=True,
           *, seed=None):
```

File: scipy/cluster/vq.py [hunk 2]
```diff
@@ -480,18 +493,20 @@ def _kpoints(data, k):
         row is one observation.
     k : int
         Number of samples to generate.
+    rng : `numpy.random.Generator` or `numpy.random.RandomState`
+        Random number generator
```
Comment (tupui) [inline scipy/cluster/vq.py:497]:
```suggestion
        Random number generator.
```

File: scipy/cluster/vq.py [hunk 3]
```diff
@@ -505,6 +520,8 @@ def _krandinit(data, k):
...
```
Comment (tupui) [inline scipy/cluster/vq.py:524]:
```suggestion
        Random number generator.
```

File: scipy/cluster/vq.py [hunk 4]
```diff
@@ -547,6 +564,8 @@ def _kpp(data, k):
...
```
Comment (tupui) [inline scipy/cluster/vq.py:568]:
```suggestion
        Random number generator.
```

File: scipy/cluster/vq.py [hunk 5]
```diff
@@ -563,15 +582,25 @@ def _kpp(data, k):
     dims = data.shape[1] if len(data.shape) > 1 else 1
     init = np.ndarray((k, dims))
 
+    # get random generator API
```
Comment (tupui) [inline scipy/cluster/vq.py:585]: We actually have this in `_util:rng_integers`. I also discovered that recently.

File: scipy/cluster/vq.py [hunk 6]
```diff
@@ -596,7 +625,7 @@ def _missing_raise():
 
 
 def kmeans2(data, k, iter=10, thresh=1e-5, minit='random',
-            missing='warn', check_finite=True):
+            missing='warn', check_finite=True, seed=None):
```
Comment (tupui) [inline scipy/cluster/vq.py:628]:
```suggestion
            missing='warn', check_finite=True, *, seed=None):
```

**Review by tupui — COMMENTED** [review]
Good addition, thanks! A few minor things on my side.

Also, I am not sure what we should use in the future to name RNG. `seed` or `random_state` or something else? @mdhaber @WarrenWeckesser (we could add this to your PR about best practice)?

File: scipy/cluster/vq.py (hunk 5 above)
Comment (AtsushiSakai) [reply scipy/cluster/vq.py]: Thanks. I used it. and I also added `_util:rng_random` and a test for it. PTAL again.

File: scipy/_lib/_util.py [hunk 7]
```diff
@@ -458,6 +458,51 @@ def __call__(self, func, iterable):
                             " form f(func, iterable)") from e
 
 
+def rng_random(gen, *, size=None, dtype=np.float64, out=None):
```
Comment (tupui) [inline scipy/_lib/_util.py:461]: I would not do this. We will soon drop support for numpy <1.17 and we should not really add more hack around it. Right now you can just use `rng.uniform` for this.

File: scipy/cluster/vq.py [hunk 8]
```diff
@@ -361,6 +363,16 @@ def kmeans(obs, k_or_guess, iter=20, thresh=1e-5, check_finite=True):
         (crashes, non-termination) if the inputs do contain infinities or NaNs.
         Default: True
 
+    seed : {None, int, `numpy.random.Generator`, `numpy.random.RandomState`}, optional
+        Seed for initializing the pseudo-random number generator.
```
Comment (tupui) [inline scipy/cluster/vq.py:367]:
```suggestion
    seed : {None, int, `numpy.random.Generator`, `numpy.random.RandomState`}, optional
        Seed for initializing the pseudo-random number generator.
```
```suggestion
    seed : {None, int, `numpy.random.Generator`,
            `numpy.random.RandomState`}, optional

        Seed for initializing the pseudo-random number generator.
```

**Review by tupui — APPROVED** [review]
LGTM with the change on the doc to respect the line length. Thanks! I will wait for the CI and merge unless you want to add something.

File: scipy/_lib/_util.py (hunk 7 above)
Comment (AtsushiSakai) [reply scipy/_lib/_util.py]: OK. I reverted it.

Comment (tupui) [reply scipy/_lib/_util.py]: I have seen thanks.

File: scipy/cluster/vq.py [hunk 9]
```diff
@@ -647,6 +669,15 @@ def kmeans2(data, k, iter=10, thresh=1e-5, minit='random',
         Disabling may give a performance gain, but may result in problems
         (crashes, non-termination) if the inputs do contain infinities or NaNs.
         Default: True
+    seed : {None, int, `numpy.random.Generator`, `numpy.random.RandomState`}, optional
+        Seed for initializing the pseudo-random number generator.
```
Comment (tupui) [inline scipy/cluster/vq.py:673]:
```suggestion
    seed : {None, int, `numpy.random.Generator`,
            `numpy.random.RandomState`}, optional

        Seed for initializing the pseudo-random number generator.
```

**Comment by tupui** [comment]
The timeout are unrelated and seen in other PRs so merging. Thanks again @AtsushiSakai

**Pull request merged**